File: src/languages/python/extract.py
"""Public API of a Python package, read with griffe (https://github.com/mkdocstrings/griffe),
the API extractor behind mkdocstrings: it resolves re-exports and aliases, `__all__`,
wildcard imports and Python's public/private conventions, statically, with a dynamic
fallback. Deprecations come from the griffe-warnings-deprecated extension
(`@warnings.deprecated` / `@typing_extensions.deprecated`).

usage: python3 extract.py <repo_root> <package_dir>
Prints "\\0JSON\\0" followed by {"symbols": [...], "warnings": [...]}.
"""
import json
import os
import sys

import griffe

MARK = "\0JSON\0"
KINDS = griffe.ParameterKind


def params_of(fn):
    out = []
    for p in fn.parameters:
        if p.name in ("self", "cls"):
            continue
        param = {"name": p.name}
        if p.annotation is not None:
            param["type"] = str(p.annotation)
        if p.default is not None or p.kind in (KINDS.var_positional, KINDS.var_keyword):
            param["optional"] = True
        if p.kind in (KINDS.var_positional, KINDS.var_keyword):
            param["rest"] = True
        if p.kind in (KINDS.keyword_only, KINDS.var_keyword):
            param["keyword"] = True
        out.append(param)
    return out


def main():
    root, pkg_rel = sys.argv[1], sys.argv[2]
    pkg_dir = os.path.normpath(os.path.join(root, pkg_rel))
    pkg_name = os.path.basename(pkg_dir)
    warnings = []
    try:
        extensions = griffe.load_extensions("griffe_warnings_deprecated")
    except Exception as e:  # noqa: BLE001
        extensions = None
        warnings.append(f"griffe-warnings-deprecated unavailable, deprecations not detected: {e}")
    package = griffe.load(
        pkg_name,
        search_paths=[os.path.dirname(pkg_dir)],
        extensions=extensions,
        resolve_aliases=True,
        resolve_external=False,
    )

    symbols = []
    seen_modules = set()

    def rel(path):
        return os.path.relpath(str(path), root) if path else None

    def walk(module, prefix):
        if module.path in seen_modules:
            return
        seen_modules.add(module.path)
        for name, member in module.members.items():
            try:
                public = member.is_public
            except Exception:  # noqa: BLE001 - unresolvable alias
                continue
            if not public or name.startswith("_"):
                continue
            try:
                kind = member.kind
            except griffe.AliasResolutionError:
                continue  # re-export of something outside the package
            if kind is griffe.Kind.MODULE and not member.is_alias:
                if any(part in ("tests", "test") or part.startswith("test_") for part in member.path.split(".")):
                    continue
                walk(member, f"{prefix}{name}.")
            elif kind is griffe.Kind.FUNCTION:
                target = member.final_target if member.is_alias else member
                sym = {"name": prefix + name, "params": params_of(target)}
                if target.returns is not None:
                    sym["returns"] = str(target.returns)
                if getattr(target, "deprecated", None):
                    sym["deprecated"] = True
                if member.is_alias:
                    target_path = target.path
                    if target_path.startswith(pkg_name + "."):
                        sym["aliasOf"] = target_path[len(pkg_name) + 1 :]
                sym["location"] = {"file": rel(target.filepath), "line": target.lineno or 0}
                symbols.append(sym)

    walk(package, "")
    sys.stdout.write(MARK + json.dumps({"symbols": symbols, "warnings": warnings}))


main()

File: src/languages/ruby/extract.rb
# Extract the public API of a Ruby lib by loading it and reflecting on its modules.
# usage: ruby extract.rb <repo_root> <entry> [namespace]
require_relative "loader"

root, entry, namespace = ARGV
roots, warnings = ApiValidatorLoader.load(root, entry, namespace)
strip_root = roots.size == 1

# Types and deprecations come from YARD (https://yardoc.org), parsed by YARD itself.
yard_lib = Dir[File.join(ENV.fetch("API_VALIDATOR_YARD", ""), "gems", "yard-*", "lib")].first
$LOAD_PATH.unshift(yard_lib) if yard_lib
require "yard"
YARD::Registry.clear
lib_dir = File.expand_path(entry, root)
YARD.parse(File.directory?(lib_dir) ? Dir[File.join(lib_dir, "**", "*.rb")] : [lib_dir], [], YARD::Logger::ERROR)

def yard_for(owner_name, meth)
  obj = YARD::Registry.at("#{owner_name}.#{meth}") || YARD::Registry.at("#{owner_name}##{meth}")
  return [{}, nil, false] unless obj
  params = obj.tags(:param).to_h { |t| [t.name.to_s, t.types] }
  ret = obj.tag(:return)&.types
  [params, ret, obj.has_tag?(:deprecated)]
end

def yard_type(types)
  types && !types.empty? ? types.join(" | ") : nil
end

# YARD type -> the shared structured type tree (src/core/model.ts), via YARD's own parser.
def type_node(t)
  case t
  when YARD::Tags::TypesExplainer::HashCollectionType
    { kind: "name", name: t.name, args: [union_node(t.key_types), union_node(t.value_types)] }
  when YARD::Tags::TypesExplainer::CollectionType # includes FixedCollectionType
    { kind: "name", name: t.name, args: t.types.map { |x| type_node(x) } }
  else
    { kind: "name", name: t.name }
  end
end

def union_node(types)
  nodes = types.map { |t| type_node(t) }
  nodes.size == 1 ? nodes.first : { kind: "union", of: nodes }
end

def yard_node(types)
  return nil if types.nil? || types.empty?
  union_node(types.flat_map { |t| YARD::Tags::TypesExplainer::Parser.parse(t) })
rescue SyntaxError, StandardError
  { kind: "unknown", text: types.join(", ") }
end

MODULE_NAME = Module.instance_method(:name)
def mod_name(m) = MODULE_NAME.bind_call(m)

symbols = []
seen = {}
walk = lambda do |mod, prefix|
  return if seen[mod]
  seen[mod] = true
  return if mod.is_a?(Class) && mod <= Exception

  names = mod.singleton_methods(false)
  names += mod.public_instance_methods(false) if !mod.is_a?(Class) && mod.singleton_class.include?(mod) # extend self
  names.uniq.sort.each do |m|
    meth = mod.method(m)
    file, line = meth.source_location
    next unless file # C methods
    params, ret, deprecated = yard_for(mod_name(mod), m)
    symbols << {
      name: [prefix, m.to_s].reject(&:empty?).join("."),
      params: meth.parameters.reject { |kind, _| kind == :block }.each_with_index.map { |(kind, name), i|
        n = (name || "arg#{i}").to_s
        {
          name: n,
          type: yard_type(params[n]),
          typeNode: yard_node(params[n]),
          optional: %i[opt rest key keyrest].include?(kind) || nil,
          rest: %i[rest keyrest].include?(kind) || nil,
          keyword: %i[key keyreq keyrest].include?(kind) || nil
        }.compact
      },
      returns: yard_type(ret),
      returnsNode: yard_node(ret),
      deprecated: deprecated || nil,
      # `alias_method :valid?, :is_valid` (or `alias`): the same definition under another name.
      aliasOf: meth.original_name == m ? nil : [prefix, meth.original_name.to_s].reject(&:empty?).join("."),
      location: { file: file.sub(%r{^#{Regexp.escape(File.expand_path(root))}/}, ""), line: line }
    }.compact
  end

  mod.constants(false).sort.each do |c|
    child = begin
      mod.const_get(c, false)
    rescue StandardError, LoadError
      nil
    end
    next unless child.is_a?(Module) && mod_name(child)&.start_with?("#{mod_name(mod)}::")
    walk.call(child, [prefix, c.to_s].reject(&:empty?).join("."))
  end
end

roots.each { |r| walk.call(r, strip_root ? "" : mod_name(r)) }
$stdout.write(ApiValidatorLoader::MARK + JSON.generate({ symbols: symbols, warnings: warnings }))
